### PR TITLE
Full convexity

### PR DESCRIPTION
**Pull request by JacquesOlivierLachaud**
# PR Description

This PR completes the module on Digital Convexity with new functions related to full convexity. In particular,  it provides an algorithm to check full convexity for arbitrary digital sets in nD, and proposes several helper classes for using full convexity in practice: one  dedicated to local geometric analysis, the other devoted to tangency and shortest paths.

# Checklist

- [x] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [x] Doxygen documentation of the code completed (classes, methods, types, members...)
- [x] Documentation module page added or updated.
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [x] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [x] All continuous integration tests pass (Travis & appveyor)

### CODE REVIEW

File: examples/geometry/volumes/geometricAnalysis3D.cpp [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+/**
+ * @file geometry/volumes/geometricAnalysis3D.cpp
```
Comment (dcoeurjo) [inline examples/geometry/volumes/geometricAnalysis3D.cpp:18]: CVXgeometricAnalysis3D (or any variations) just to be more precise ?

File: examples/geometry/volumes/geometricThinning3D.cpp [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+/**
+ * @file geometry/volumes/geometricThinning3D.cpp
```
Comment (dcoeurjo) [inline examples/geometry/volumes/geometricThinning3D.cpp:18]: same here ? cvx in the name?

File: examples/geometry/volumes/lutFullConvexity2D.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+/**
+ * @file topology/generateSimplicityTables2D.cpp
```
Comment (dcoeurjo) [inline examples/geometry/volumes/lutFullConvexity2D.cpp:18]: wrong file name in the @file

File: examples/geometry/volumes/lutFullConvexity2D.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+ * @ingroup Examples
+ * @author Jacques-Olivier Lachaud (\c jacques-olivier.lachaud@univ-savoie.fr )
+ * Laboratory of Mathematics (CNRS, UMR 5127), University of Savoie, France
+ *
+ * @date 2012/06/22
+ *
+ * An example file named generateSimplicityTables2D. Creates precomputed
```
Comment (dcoeurjo) [inline examples/geometry/volumes/lutFullConvexity2D.cpp:25]: fixme as well

File: examples/geometry/volumes/shortestPaths3D.cpp [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+/**
+ * @file geometry/volumes/shortestPaths3D.cpp
```
Comment (dcoeurjo) [inline examples/geometry/volumes/shortestPaths3D.cpp:18]: same ? cvx keyword?

Comment (dcoeurjo) [reply examples/geometry/volumes/shortestPaths3D.cpp:18]: tangencyShortestPaths3D ?

File: examples/geometry/volumes/simpleness-convexity-2d.cpp [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+/**
+ * @file topology/generateSimplicityTables2D.cpp
```
Comment (dcoeurjo) [inline examples/geometry/volumes/simpleness-convexity-2d.cpp:18]: wrong file?

Comment (dcoeurjo) [reply examples/geometry/volumes/simpleness-convexity-2d.cpp:18]: `simpleness` is a bit odd to me

File: src/DGtal/base/TimeStampMemoizer.h [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+//////////////////////////////////////////////////////////////////////////////
+
+namespace DGtal
+{
+
+  /////////////////////////////////////////////////////////////////////////////
+  // template class TimeStampMemoizer
+  /**
+   * Description of template class 'TimeStampMemoizer' <p> \brief Aim:
+   * A generic class to store costly computations. A computation is an
```
Comment (dcoeurjo) [inline src/DGtal/base/TimeStampMemoizer.h:55]: is that a cache mechanism?

Comment (dcoeurjo) [reply src/DGtal/base/TimeStampMemoizer.h:55]: more than stamping thing?

Comment (dcoeurjo) [reply src/DGtal/base/TimeStampMemoizer.h:55]: oh no I see.. a kind of key/value buffered container + timestamp

File: src/DGtal/base/TimeStampMemoizer.h [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    }
+
+    /// @return the maximum number of items that can be memoized.
+    Size maxSize() const
+    {
+      return myMaxSize;
+    }
+
+    /// @return the number of successful hits in the memoizer
+    Size hits() const
```
Comment (dcoeurjo) [inline src/DGtal/base/TimeStampMemoizer.h:145]: what `hit` means in this context?

File: src/DGtal/base/TimeStampMemoizer.h (hunk 7 above)
Comment (dcoeurjo) [reply src/DGtal/base/TimeStampMemoizer.h:55]: but I still don't get the "costly part..."

File: src/DGtal/base/TimeStampMemoizer.h (hunk 8 above)
Comment (dcoeurjo) [reply src/DGtal/base/TimeStampMemoizer.h:145]: oh.. hit as in the memory cache access policy I guess

File: src/DGtal/geometry/doc/moduleDigitalConvexity.dox [hunk 9]
```diff
@@ -6,27 +6,38 @@ namespace DGtal {
...
 @writers Jacques-Olivier Lachaud
 
 @since 1.1
 
 Part of the \ref packageGeometry.
   
-This part of the manual describes how to define and check digital convexity, 
-in such a way that digital convex set are digitally connected.
+This part of the manual describes tools associated to a new definition
+of digital convexity, called full convexity
```
Comment (dcoeurjo) [inline src/DGtal/geometry/doc/moduleDigitalConvexity.dox:18]:
```suggestion
of digital convexity, called the full convexity
```

File: src/DGtal/geometry/doc/moduleDigitalConvexity.dox [hunk 10]
```diff
@@ -6,27 +6,38 @@ namespace DGtal {
...
+\cite lachaud_dgmm_2021 . This new definition solves many problems
+related to the usual definition of digital convexity, like possible
+non connectedness or non simple connectedness, while encompassing its
+desirable features. Fully convex sets are digitally convex, but are
+also connected and simply connected. They have a morphological
+characterisation, which induces a simple convexity test
+algorithm. Arithmetic planes are fully convex too.
```
Comment (dcoeurjo) [inline src/DGtal/geometry/doc/moduleDigitalConvexity.dox:25]:
```suggestion
algorithm. For example, arithmetic planes are fully convex.
```

File: src/DGtal/geometry/doc/moduleDigitalConvexity.dox [hunk 11]
```diff
@@ -219,20 +242,55 @@ using namespace DGtal::Z3i;
...
-bool ok  = simplex_cover.subset( point_cover ); // one way to do it
-bool ok2 = dconv.isFullyConvex ( fat_simplex ); // the other way to do it
+// Checks that C^d[ Conv(Z) ] is a subset of C^d[ Z ], i.e. Z is fully convex.
+bool ok  = simplex_cover.subset( point_cover ); // true
 \endcode
 
+@subsection dgtal_dconvexity_sec24 Other convexity services, like digital subconvexity (or tangency)
+
+Morphological services (since 1.3):
+- DigitalConvexity::makePolytope( const PointRange& X, bool safe ) const builds the tightiest polytope enclosing the digital set \a X
```
Comment (dcoeurjo) [inline src/DGtal/geometry/doc/moduleDigitalConvexity.dox:280]:
```suggestion
- DigitalConvexity::makePolytope( const PointRange& X, bool safe ) const builds the tightest polytope enclosing the digital set \a X
```

File: src/DGtal/geometry/doc/moduleDigitalConvexity.dox [hunk 12]
```diff
@@ -219,20 +242,55 @@ using namespace DGtal::Z3i;
...
+- DigitalConvexity::is0Convex( const PointRange& X, bool safe ) const checks the usual digital convexity (H-convexity or 0-convexity) of the digital set \a X
+- DigitalConvexity::U( Dimension i, const PointRange& X ) const performs the digital Minkowski sum of \a X along direction \a i
+
+The following snippet shows that a 4D ball is indeed fully convex.
+
+@snippet testFullConvexity.cpp nD-full-convexity
+
+
+@note This method for checking full convexity is slightly slower than
+the method using Minkowski sum on the polytope contraints, but it
```
Comment (dcoeurjo) [inline src/DGtal/geometry/doc/moduleDigitalConvexity.dox:291]:
```suggestion
the method using the Minkowski sum on the polytope contraints, but it
```

File: src/DGtal/geometry/doc/moduleDigitalConvexity.dox [hunk 13]
```diff
@@ -249,8 +307,77 @@ Lattice cell geometry services:
...
-- DigitalConvexity::isKSubconvex tells if a given polytope is k-subconvex to some cell cover
+- DigitalConvexity::isKSubconvex tells if a given polytope is k-subconvex to some cell cover (i.e. k-tangent)
+- DigitalConvexity::isKSubconvex( const Point&, const Point&, const CellGeometry&, const Dimension ) const tells if the Euclidean straight segment \f$ \lbrack a,b \rbrack \f$ is k-subconvex (i.e. k-tangent) to the given cell cover.
 - DigitalConvexity::isFullySubconvex tells if a given polytope is fully subconvex to some cell cover
+- DigitalConvexity::isFullySubconvex( const Point&, const Point&, const CellGeometry& ) const tells if the Euclidean straight segment \f$ \lbrack a,b \rbrack \f$ is fully k-subconvex (i.e. tangent) to the given cell cover.
+
+@subsection dgtal_dconvexity_sec25 Ehrhart polynomial of a lattice polytope
+
+Any lattice polytope has a unique Ehrhart polynomial that encodes the
+relationship between the volume of a polytope and the number of
```
Comment (dcoeurjo) [inline src/DGtal/geometry/doc/moduleDigitalConvexity.dox:318]:
```suggestion
relationship between the volume of the polytope and the number of
```

File: src/DGtal/geometry/doc/moduleDigitalConvexityApplications.dox [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+Part of the \ref packageGeometry.
+  
+This part of the manual describes some applications of a new
+definition of digital convexity, called full convexity \cite lachaud_dgmm_2021 .
+See \ref moduleDigitalConvexity for further details on full convexity.
+
+[TOC]
+
+The following programs are related to this documentation:
+geometry/volumes/geometricAnalysis3D.cpp ,
```
Comment (dcoeurjo) [inline src/DGtal/geometry/doc/moduleDigitalConvexityApplications.dox:24]: renaming suggestions

File: src/DGtal/geometry/doc/moduleDigitalConvexityApplications.dox [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+\} \f$, it holds that \f$ \{ x_i, x_{i+1} \} \f$ is tangent to \f$ X \f$. 
+
+We can embed in the Euclidean space the path \f$ \gamma \f$ simply by
+joining its consecutive points by Euclidean straight line
+segments. Its \a length is then just the Euclidean length of its embedding.
+
+The cotangency relations define a graph on \f$ X \f$, which can be
+weighted by the length of the Euclidean segments joining cotangent
+points. Shortest paths on this graph are of course \a path in the
+sense above.
```
Comment (dcoeurjo) [inline src/DGtal/geometry/doc/moduleDigitalConvexityApplications.dox:142]:
```suggestion
above mentioned sense.
```

File: src/DGtal/geometry/volumes/CellGeometry.ih [hunk 16]
```diff
@@ -80,6 +81,122 @@ init( const KSpace & K,
...
+                  kc[ j ] = 2 * ( a[ j ] + q );
+                  if ( r < 0 )      kc[ j ] -= 1;
+                  else if ( r > 0 ) kc[ j ] += 1;
+                }
+            }
+          addCellsTouchingCell( myK.uCell( kc ) );
+        }
+    }
+  if ( a != b ) addCellsTouchingPoint( b );
+  // Put too many cells in some cases.
```
Comment (dcoeurjo) [inline src/DGtal/geometry/volumes/CellGeometry.ih:166]: cleanup ? or add more comments... (maybe an issue on GitHub)

**Review by dcoeurjo — COMMENTED** [review]
Just a first batch of comments ;)

File: src/DGtal/geometry/volumes/NeighborhoodConvexityAnalyzer.h [hunk 17]
```diff
@@ -0,0 +1,623 @@
...
+#include "DGtal/topology/CCellularGridSpaceND.h"
+#include "DGtal/topology/KhalimskySpaceND.h"
+#include "DGtal/geometry/volumes/DigitalConvexity.h"
+//////////////////////////////////////////////////////////////////////////////
+
+namespace DGtal
+{
+  namespace detail {
+    
+    template< typename T >
```
[the remark this answers is not in the document]
Comment (dcoeurjo) [reply src/DGtal/geometry/volumes/NeighborhoodConvexityAnalyzer.h:71]: a new `base/ConstExpressions.h` for factorize such things ?
(+ in add them in  the `functions::` namespace)

File: src/DGtal/geometry/volumes/ConvexityHelper.h [hunk 18]
```diff
@@ -211,6 +293,52 @@ namespace DGtal
...
+    ///
+    /// @param input_points a range of distinct points.
+    ///
+    /// @return the tightiest bounded lattice polytope
+    /// (i.e. H-representation) including the given range of points,
+    /// or an empty polytope if the given range of points was not full
+    /// dimensional and dimension was greater than 3.
+    static
+    LatticePolytope
+    computeDegeneratedLatticePolytope( std::vector< Point > input_points );
```
[the remark this answers is not in the document]
Comment (dcoeurjo) [reply src/DGtal/geometry/volumes/ConvexityHelper.h:342]: I see, but for all non-degenerated cases, the deep copy will be performed and you might not need it, right?

**Review by dcoeurjo — COMMENTED** [review]
Just a couple of minor questions. Otherwise, no more comments, thx

File: src/DGtal/geometry/volumes/NeighborhoodConvexityAnalyzer.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/geometry/volumes/NeighborhoodConvexityAnalyzer.h]: Done

File: src/DGtal/geometry/volumes/ConvexityHelper.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/geometry/volumes/ConvexityHelper.h]: Done  with inout argument passing.

**Comment by JacquesOlivierLachaud** [comment]
Should be  fine now.

**Comment by dcoeurjo** [comment]
Great, thx

**Review by dcoeurjo — APPROVED** [review]
All good

**Pull request merged**